platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,304 @@
+
+   This assumes that the broker is accessible from CI. There may be some work involved in configuring the the broker to allow access depending on where it's hosted.
+4. Publishing the pacts should trigger a webhook in the broker that invokes a Rake task in `vets-api`:
+   ```
+   bundle exec rake pact:verify
+   ```
+   This task pulls all relevant pacts from the broker and verifies that the expected responses match the actual responses when running the requests through a running instance of the API. For any states specified in the pacts, the API will set up any matching states already defined in `provider_states.rb`. If there are no matching states, the verification will fail.
+   
+   [Pact Broker webhooks can be created using the CLI](https://github.com/pact-foundation/pact_broker-client/#create-webhook), whether it's from the Ruby gem or NPM package. **The webhook should just check for `contract_content_changed` because the broker will automatically pass any unmodified contracts that previously passed verification."**
+   
+   While the pacts are being verified, the `vets-website` build can proceed to E2E tests.
+
+   **If the pacts are coming from a feature branch and the API has not been updated to match the new expectations, verification should be expected to fail.** In that case, once the pacts have been published, the BE developers working on updating the API can run the following task during development to verify their changes:
+   ```
+   bundle exec rake pact:verify:at[http://build-box/MyConsumerBuild/latestSuccessful/artifacts/my_consumer-my_provider.json]
+   ```
+   The actual URL should point to the contract that was published from the `vets-website` feature branch. After the `vets-api` feature branch is merged to master, that pact should be able to pass verification. 
+5. **The verification results are published back to the broker** with the version of the API that verified the pacts. Again, we can use the commit hash to version, but this time it's from `vets-api`.
+   The broker maintains a matrix matching the consumer version and provider version.
+6. **The `vets-website` build runs a check to see if it can be deployed.**
+   ```
+   pact-broker can-i-deploy --broker-base-url=BROKER_BASE_URL
+   ```
+   This check is dependent on the published verification results.
+
+   **If verification was successful, the CI pipeline will proceed to deploy. If it failed, the pipeline will stop.**
+
+   We're assuming that the CI pipeline can wait for the asynchronous verification task to run and conditionally handle the output of that task.
```

Comment:
is this a valid assumption?

if not, is there a fallback plan?

Reply:
From reading Circle's documentation, I think it's supported, but I haven't wrapped my head around it yet.

The fallback plan would be to include running the `vets-api` verification task in the same `vets-website` CI pipeline but in a job following the publishing of the pacts. It can either pull from the broker (just to ensure that it's all there; I think this option is better) or from the artifacts generated by the publishing step in the Circle pipeline (if we want to store the pacts as a build artifact) and would avoid async concerns (and seems a lot easier).

The verification job triggered by `vets-website` pushes wouldn't be under the `vets-api` organization in Circle, but I think we can live with that.

The verifications triggered by `vets-api` pushes would still remain under that org, so I don't see this fallback approach affecting it.

I didn't mention this alternative originally, but it seems worth adding now that I've written it out.